source/posix/socket.c

```diff
@@ -1254,7 +1254,94 @@ int aws_socket_set_options(struct aws_socket *socket, const struct aws_socket_op
             socket->io_handle.data.fd,
             errno_value);
     }
-
+    size_t network_interface_length = 0;
+    if (aws_secure_strlen(options->network_interface_name, AWS_NETWORK_INTERFACE_NAME_MAX, &network_interface_length)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name max length must be %d length and NULL terminated",
+            (void *)socket,
+            socket->io_handle.data.fd,
+            AWS_NETWORK_INTERFACE_NAME_MAX);
+        return aws_raise_error(AWS_IO_SOCKET_INVALID_OPTIONS);
+    }
+    if (network_interface_length != 0) {
+#if defined(AWS_OS_APPLE)
```

Comment:
instead of this:
```
if APPLE: use IP_BOUND_IF
elif: LINUX SO_BINDTODEVICE`
else: bork
```

seems more flexible (in a world with 10,000 Unix variants) to do something like this:
```
if defined(SO_BINDTODEVICE): use SO_BINDTODEVICE
elif defined(IP_BOUND_IF): use IP_BOUND_IF
else: bork
```

Reply:
Yeah, it's more flexible, but I think it's better to be exact about where we want to support this in a world of 10,000 Unix variants and this feature requiring Linux kernel >= 5.7 or root permissions, etc. If we make it more flexible and try to support all Unix variants instead of just Linux and macOS, it will increase our support scope, which I am not sure if we want to do unless someone asks for it.

Reply:
I have updated the ifdef's